syntax = "proto3";

package dfs_service;

service DFSService {

    // Add your service calls here

    // 1. REQUIRED (Parts 1 & 2): A method to store files on the server
    rpc Store(stream StoreRequest) returns (StoreResponse);
    // 2. REQUIRED (Parts 1 & 2): A method to fetch files from the server
    rpc Fetch(FetchRequest) returns (stream FetchResponse);
    // 3. REQUIRED (Parts 1 & 2): A method to list all files on the server
    rpc List(ListRequest) returns (ListResponse);
    // 4. REQUIRED (Parts 1 & 2): A method to get the status of a file on the server
    rpc Stat(StatRequest) returns(StatResponse);
    // 5. REQUIRED (Part 2 only): A method to request a write lock from the server
    rpc RequestWriteLock(WriteLockRequest) returns (WriteLockResponse);
    // 6. REQUIRED (Part 2 only): A method named CallbackList to handle asynchronous file listing requests
    //                            from a client. This method should return a listing of files along with their
    //                            attribute information. The expected attribute information should include name,
    //                            size, modified time, and creation time.

    // 7. REQUIRED (Part 2 only): A method to delete a file from the server
    rpc CallbackList(FileRequest) returns (FileList);
    // 8. Any other methods you deem necessary to complete the tasks of this assignment
    rpc Delete(DeleteRequest) returns (DeleteResponse);

}

// Add your message types here

message FileInfo {
    string filename = 1;
    int64 size = 2;
    int64 mtime = 3;
    uint32 crc = 4;  
}

message StoreRequest {
    bytes data = 1;
}

message StoreResponse {
    int32 status = 1;
    string messageText = 2;
    string filename = 3; 
    int64 mtime = 4;
}

message FetchRequest {
    string filename = 1;
}

message FetchResponse {
    bytes data = 1;
}

message DeleteRequest {
    string filename = 1;
}

message DeleteResponse {
    string message = 1;
    int32 status = 2; 
}

message ListRequest {
}

message ListResponse {
    repeated FileInfo files = 1;
}

message StatRequest {
    string filename = 1;
}

message StatResponse {
    FileInfo file_info = 1;
}

message WriteLockRequest {
    string filename = 1;
    string client_id = 2;
}

message WriteLockResponse {
    bool granted = 1;   
}

message FileRequest {
    string client_id = 1;
    string name = 2;
}

message FileList {
    repeated FileInfo file_info = 1;
}